clang/test/Sema/flexible-array-in-union.c

```diff
@@ -1,13 +1,158 @@
-// RUN: %clang_cc1 %s -verify=c -fsyntax-only
-// RUN: %clang_cc1 %s -verify -fsyntax-only -x c++
-// RUN: %clang_cc1 %s -verify -fsyntax-only -fms-compatibility
-// RUN: %clang_cc1 %s -verify -fsyntax-only -fms-compatibility -x c++
+// RUN: %clang_cc1 %s -verify=stock,c -fsyntax-only
+// RUN: %clang_cc1 %s -verify=stock,cpp -fsyntax-only -x c++
+// RUN: %clang_cc1 %s -verify=stock,cpp -fsyntax-only -fms-compatibility -x c++
+// RUN: %clang_cc1 %s -verify=stock,c,gnu -fsyntax-only -Wgnu-flexible-array-union-member -Wgnu-empty-struct
+// RUN: %clang_cc1 %s -verify=stock,c,microsoft -fsyntax-only -fms-compatibility -Wmicrosoft
 
 // The test checks that an attempt to initialize union with flexible array
 // member with an initializer list doesn't crash clang.
 
 
-union { char x[]; } r = {0}; // c-error {{flexible array member 'x' in a union is not allowed}}
+union { char x[]; } r = {0}; /* gnu-warning {{flexible array member 'x' in a union is a GNU extension}}
+                                microsoft-warning {{flexible array member 'x' in a union is a Microsoft extension}}
+                              */
+struct _name1 {
+  int a;
+  union {
+    int b;
+    char x[]; /* gnu-warning {{flexible array member 'x' in a union is a GNU extension}}
+                 microsoft-warning {{flexible array member 'x' in a union is a Microsoft extension}}
+               */
+  };
+} name1 = {
+  10,
+  42,        /* initializes "b" */
+};
 
-// expected-no-diagnostics
+struct _name1i {
+  int a;
+  union {
+    int b;
+    char x[]; /* gnu-warning {{flexible array member 'x' in a union is a GNU extension}}
+                 microsoft-warning {{flexible array member 'x' in a union is a Microsoft extension}}
+               */
+  };
+} name1i = {
+  .a = 10,
+  .b = 42,
+};
+
+/* Initialization of flexible array in a union is never allowed. */
+struct _name2 {
+  int a;
+  union {
+    int b;
+    char x[]; /* gnu-warning {{flexible array member 'x' in a union is a GNU extension}}
+                 microsoft-warning {{flexible array member 'x' in a union is a Microsoft extension}}
+                 stock-note {{initialized flexible array member 'x' is here}}
+               */
+  };
+} name2 = {
+  12,
+  13,
+  { 'c' },   /* stock-error {{initialization of flexible array member is not allowed}} */
```

Comment:
I was surprised by this too, especially since using named initializers does produce a useful warning. Improving this message seems out of scope for this PR, though, yes?

Reply:
I think this is more than just a confusing error message. Consider `struct S { int a; union { int b; char x[]; }; int c; } s = {1, 2, 3};`. If `x` is given type `char[0]`, or `char[1]`, or `char[2]`, the `3` initializer applies to `s.c`. Yet inexplicably, if `x` has type `char[]`, the `3` initializer applies to `s.x` and any initializer is an error? That looks like a bug. We can even turn this into wrong-code: `struct S { int a; union { int b; char x[]; }; int c; int d; } s = {1, 2, {}, 3};`. This should initialise `a`, `b`, `c`, `d`, to `1`, `2`, `0`, `3`, but instead initialises them to `1`, `2`, `3`, `0`.

Reply:
Yeah, I see what you mean. It looks like (when not using named initializers) a union initializer is applied to the first member. (i.e. when the type is anything other than a flexible array.) When it's a flex array, it processes it but only allows it to be zero initialized. Regardless, here's the results for unnamed initializers:

```
struct { int a; union { int b; short x; }; int c; int d; } h = {1, 2, {}, 3};
// @h = global %struct.anon{{.*}} { i32 1, %union.anon{{.*}} { i32 2 }, i32 0, i32 3 }
struct { int a; union { int b; short x[0]; }; int c; int d; } h0 = {1, 2, {}, 3};
// @h0 = global %struct.anon{{.*}} { i32 1, %union.anon{{.*}} { i32 2 }, i32 0, i32 3 }
struct { int a; union { int b; short x[1]; }; int c; int d; } h1 = {1, 2, {}, 3};
// @h1 = global %struct.anon{{.*}} { i32 1, %union.anon{{.*}} { i32 2 }, i32 0, i32 3 }
struct { int a; union { int b; short x[]; }; int c; int d; } hf = {1, 2, {}, 3};
// @hf = global %struct.anon{{.*}} { i32 1, %union.anon{{.*}} { i32 2 }, i32 3, i32 0 }
```

`hf` has clearly gone weird.

And here's what happens when using named initializers:

```
struct { int a; union { int b; short x; }; int c; int d; } hn = {.a = 1, .x = 2, .c = 3};
// @hn = global { i32, { i16, [2 x i8] }, i32, i32 } { i32 1, { i16, [2 x i8] } { i16 2, [2 x i8] undef }, i32 3, i32 0 }
struct { int a; union { int b; short x[0]; }; int c; int d; } hn0 = {.a = 1, .x = {2}, .c = 3};
// @hn0 = global { i32, { [0 x i16], [4 x i8] }, i32, i32 } { i32 1, { [0 x i16], [4 x i8] } { [0 x i16] zeroinitializer, [4 x i8] undef }, i32 3, i32 0 }
struct { int a; union { int b; short x[1]; }; int c; int d; } hn1 = {.a = 1, .x = {2}, .c = 3};
// @hn1 = global { i32, { [1 x i16], [2 x i8] }, i32, i32 } { i32 1, { [1 x i16], [2 x i8] } { [1 x i16] [i16 2], [2 x i8] undef }, i32 3, i32 0 }
```

The initialization of `x` in `hn0` just disappears, but it does generate the warning `excess elements in array initializer`, which I guess makes sense if it's not treating it as a fake flexible array?

Notably tttempting to initialize the flexible array by name will fail to build:
```
/srv/code/llvm-project/clang/test/CodeGen/flexible-array-init.c:48:82: error: initialization of flexible array member is not allowed                                                                    
   48 | struct { int a; union { int b; short x[]; }; int c; int d; } hnf = {.a = 1, .x = {2}, .c = 3 };                                                                                                  
      |                                                                                  ^          
```

Named initializers appears to be working as intended. But it seems that unnamed initializer parsing isn't skipping a flex array like it would for other types?

Reply:
A quick attempt suggests
```
diff --git a/clang/lib/Sema/SemaInit.cpp b/clang/lib/Sema/SemaInit.cpp
index 011deed7a9a9..dcb3a7db9812 100644
--- a/clang/lib/Sema/SemaInit.cpp
+++ b/clang/lib/Sema/SemaInit.cpp
@@ -2436,7 +2436,8 @@ void InitListChecker::CheckStructUnionTypes(
     }
   }
 
-  if (Field == FieldEnd || !Field->getType()->isIncompleteArrayType() ||
+  if (Field == FieldEnd || RD->isUnion() ||
+      !Field->getType()->isIncompleteArrayType() ||
       Index >= IList->getNumInits())
     return;
 
```
but it might need more tweaking for other test cases. Is this something you want to fix up and include in this PR, or is this something you would prefer to see done as a separate PR first?

Reply:
Prior to this PR, it wasn't possible to use a flex array in a union, so I'll include it here and recheck the entire test suite. Thanks for finding this! I had found the function, but hadn't yet figured out how the position was being walked. :)

Reply:
This seems to work:

```diff
diff --git a/clang/lib/Sema/SemaInit.cpp b/clang/lib/Sema/SemaInit.cpp
index 011deed7a9a9..79cf2eed46fe 100644
--- a/clang/lib/Sema/SemaInit.cpp
+++ b/clang/lib/Sema/SemaInit.cpp
@@ -2437,6 +2437,7 @@ void InitListChecker::CheckStructUnionTypes(
   }
 
   if (Field == FieldEnd || !Field->getType()->isIncompleteArrayType() ||
+      (RD->isUnion() && Field != RD->field_begin()) ||
       Index >= IList->getNumInits())
     return;
 
```